### PR TITLE
[clang][ASTImporter] Fix import of template parameter default values.

### PR DESCRIPTION
**Pull request by balazske**
Default values of template parameters (non-type, type, template) were not correctly handled in the "inherited" case. This occurs if the first declaration contains the default value but a next one not. The default value is "inherited" from the first.

In ASTImporter it was only possible to set the inherited status after the template object was created with the template parameters that were imported without handling the inherited case. The import function of the template parameter contains not enough information (previous declaration) to set the inherited-from status. After the template was created, default value of the parameters that should be inherited is reset to inherited mode.

### CODE REVIEW

File: clang/lib/AST/ASTImporter.cpp [hunk 1]
```diff
@@ -359,6 +359,31 @@ namespace clang {
...
+      if (auto *ParmT = dyn_cast<TemplateParmDeclT>(RecentParm)) {
+        if (ParmT->hasDefaultArgument()) {
+          auto *P = cast<TemplateParmDeclT>(NewParm);
+          P->removeDefaultArgument();
+          P->setInheritedDefaultArgument(Importer.ToContext, ParmT);
+        }
+      }
+    }
+
+    void updateTemplateParametersInheritedFrom(
```
Comment (NagyDonat) [inline clang/lib/AST/ASTImporter.cpp:374]:
```suggestion
  /// Update the parameter list `NewParams` of a template declaration by
  /// "inheriting" default argument values from `RecentParams`, which is
  /// the parameter list of an earlier declaration of the same template.
  /// (Note that "inheriting" default argument values this way is
  /// completely unrelated to the usual object-oriented "inheritance"
  /// relationship between classes.)
  ///
  /// Note that the standard specifies that the same parameter cannot be
  /// given default arguments twice in the same scope; but in our case
  /// the parameter list `NewParams` is freshly imported from a
  /// different TU, so it's possible that both `RecentParams` and
  /// `NewParams` specify a default argument for the same parameter.
  /// In this case, the code will use the default argument taken from
  /// `RecentParams`.
  /// FIXME: Report an error if the two default arguments are different.
    void updateTemplateParametersInheritedFrom(
```
I felt that it is important to explain that the "Inherited" within the name of this function is unrelated to the usual object-oriented "inheritance" between classes; and as I started to write this comment, I also ended up documenting the goals of this function. (I hope that I was correct -- please fix this comment block if I misunderstood something!)

**Review by NagyDonat — APPROVED** [review]
LGTM, I'm very grateful that you can understand and resolve these tricky `ASTImporter` issues.

File: clang/lib/AST/ASTImporter.cpp (hunk 1 above)
Comment (balazske) [reply clang/lib/AST/ASTImporter.cpp:397]: Comment was added but different because the case is not exactly as described here.

Comment (NagyDonat) [reply clang/lib/AST/ASTImporter.cpp:397]: Thanks for the clarifications!

**Review by NagyDonat — APPROVED** [review]
